test/CMakeLists.txt

```diff
@@ -5,27 +5,27 @@ find_package(catkin REQUIRED COMPONENTS class_loader)
 
 include_directories(${catkin_INCLUDE_DIRS})
 
-add_library(${PROJECT_NAME}_TestPlugins1 EXCLUDE_FROM_ALL plugins1.cpp)
```

Comment:
It's in the `test` directory, do we need to touch this code at all?

Reply:
Undid the change and rebased. I only wanted to highlight the fact that the test libraries need to be build as shared libs otherwise the unit tests will fail. But this goes back to the point I made above and if we enforce BUILD_SHARED_LIBS for `class_loader` then we don't need this.

> > Added target exports for `class_loader`.
> > Made `TestPlugins1` and `TestPlugins2` libraries shared.
> > I want to highlight something regarding `class_loader`. [class_loader](http://wiki.ros.org/class_loader) is a ROS-independent package that allows dynamic loading of exported C++ classes during runtime from a runtime library and creating objects of those classes. We also have
> > [pluginlib](wiki.ros.org/pluginlib), which uses `class_loader` and does the same but for ROS packages.
> > Let's take the following example - we have `class_loader`, `myApp` and `myPlugin`. We want to register `myPlugin` as a plugin with `class_loader` so that `myApp` can load this plugin without being linked to it.
> > `myPlugin` must link against `class_loader` and call a macro in order to register as a plugin. When `myPlugin` is loaded by `class_loader`, `class_loader` populates a static variable that points to all the classes in that plugin. As a result, `myApp` can now instatitate these classes without being linked to `myPlugin`.
> > However there is an issue when linking `myPlugin` statically with `class_loader` and linking `myApp` statically with `class_loader`. I am not entirely sure I completely understand why this happens, but what I see is that when `myPlugin` is loaded by `class_loader` and the static variable is populated, `myApp` does not see this. Instead when `myApp` tries to instantiate a class from `myPlugin`, the same static variable in `class_loader` is not initialized yet, or rather zero initialized. I assume this has something to do with the fact that both `myPlugin` and `myApp` are statically linked against `class_loader` so they each use their own "instance" of `class_loader`? Does that make sense?
> > Because of this, I think we must compile `class_loader` with shared libraries.
> > We can see this behavior when running the unit tests in `class_loader`. The above scenario is pretty much what goes on in the tests. Linking statically against `class_loader` will cause the tests to fail. Building with shared libs and the tests pass.
> 
> I believe this analysis is correct, it should be perhaps mentioned in the documentation page later on in Hunter Sphinx for the affected packages.
> 
> It is also quite tricky to work with this plugin system in Hunter ecosystem, because the `libclass_loader.so` has to be present somewhere in the system during runtime. This could come from the actual ROS distribution from `/opt/ros/melodic`. It's not ideal because there could be version mismatch, but should do in my opinion.

Yes, I agree. I will document this in the sphinx doc for `class_loader` when I do the pull request in Hunter. I also think we should set the cmake args for BUILD_SHARED_LIBS=ON in the hunter.cmake for the package to have this as the default behavior.